lib/Dialect/TritonGPU/IR/Dialect.cpp

```diff
@@ -1551,22 +1551,23 @@ AMDMfmaEncodingAttr::getSizePerThreadForOperands(unsigned opIdx) const {
   }
 }
 
-SmallVector<unsigned>
-AMDMfmaEncodingAttr::getShapePerCTATileForDotOperands(ArrayRef<int64_t> shape,
-                                                      int opIdx) const {
-  assert(getMDim() == 32);
+SmallVector<unsigned> AMDMfmaEncodingAttr::getShapePerCTATileForDotOperands(
```

Comment:
Can you elaborate on the use case of this function? It's a little confusing what should be returned from the name of this function.

Reply:
If I understand NV path correctly, it returns size of tile processed by one matrix instruction.
I.e. [`<warps per dim> * <instruction size over dim>`], for example if `warpsPerCTA=[2,4]` and instruction MNK shape is `[16,16, 8]`, operand 0 should return `[2 * 16, 8]`, operand 1 should return `[8, 4*16]`

This function is used in ttg->llir conversions for various `ConvertLayoutOp` with src/dst mfma dot op layout.

Interestingly in our current design this function can not be called:
- some `mfma -> mfma dot op` variants are covered by shortcut, no allocation or instructions are emitted
- other `distrubuted -> mfma dot op` variants are decomposed by `DecomposeUnsupportedConversions` pass
- `mfma dot op -> distributed` variant is hard to get from triton source code (I have no idea how to get it this way), if artificially created, if fails in several other places during compilation

I actually hoped that someone from our team saw this function already and can confirm it's semantics, but looks it is better to ask someone from NV developers.

Reply:
Why is it only used between distributed layouts? How about `local_load`?

Reply:
`local_load` is processed by separate module in `SharedToDotOperandMFMA.cpp`, which do not use this function.

Reply:
Ok. Then what is this PR fixing?

Reply:
This PR is confusing, let me add more details, maybe I should break it in several PRs.

Let me give more details

Test that fails is `test_op_varlen_mqa_fwd` from https://github.com/ROCm/triton/blob/triton-mlir/python/perf-kernels/flash-attention.py.

This test generates `layout_conversion <mfma16> -> <dot op mfma16>`, which falls into shortcut category.

This layout_conversion triggers following assert in `getShapePerCTATileForDotOperands` : https://github.com/openai/triton/pull/3790/files#diff-c05cf3aed297bf0c5f1296cc40c522b00fb300c7a4340a1f6be5b0bbe2c42039L1557 while computing scratch pad size needed for this conversion.

Later in code results of `getShapePerCTATileForDotOperands` call are dropped (because this conversion is processed by shortcut and no buffer is needed).

So actual fixes of the problem are:
1. Remove assert  in `getShapePerCTATileForDotOperands` and do not care about result, because results are unused anyway;
2. Do not call this function in the first place (this is implemented here: https://github.com/openai/triton/pull/3790/files#diff-c8636c7e4e8a1713c11e249836ccf2fe132ffc8fd85ad7054582ecad544e4a26R69)

Along with fix for the test, I've decided to fix function itself, to align it with NV MMA version, now I think it is better to make a separate PR.

Reply:
I think we still need this function for mfma --> dotOp that does not fall into the shortcut, as in the naive implementation of FA-f8 support. 
@jtang10 can confirm.

Does the fix for pattern benefit necessary to pass the lit tests of this PR? If so, we kind of need it in the same PR. (especially in the case that this PR is already waiting for #3676 to fix the kWidth print)

Reply:
> Does the fix for pattern benefit necessary to pass the lit tests of this PR? If so, we kind of need it in the same PR.

Yes, this is required for lit test.
For some reason in lit tests "common" patterns are applied before AMD specific, because they have same priority and llvm is free to choose any pattern that applied to given operation.

I am actually surprised we did not see this error before.